crates/db/src/traits.rs

```diff
@@ -150,6 +150,11 @@ pub trait ClientStateDatabase {
 
     /// Gets a state checkpoint at a previously written index, if it exists.
     fn get_state_checkpoint(&self, idx: u64) -> DbResult<Option<ClientState>>;
+
+    fn get_last_client_state_idx(&self) -> DbResult<u64>;
+
+    /// Get client state at idx
+    fn get_state_at(&self, idx: u64) -> DbResult<Option<ClientState>>;
```

Comment:
No!  You don't want to make the database do this, it creates circular dependencies.  The consumer should be doing the fetches itself and reconstruct the client state itself.

Reply:
Reconstructing it in `rpc_server.rs` is fine? Please comment on the next commit that I'm gonna make.